c10/core/DispatchKeySet.h

```diff
@@ -74,31 +252,63 @@ class DispatchKeySet final {
     return DispatchKeySet(repr_ | other.repr_);
   }
   // Perform set intersection
-  DispatchKeySet operator&(DispatchKeySet other) const {
+  constexpr DispatchKeySet operator&(DispatchKeySet other) const {
     return DispatchKeySet(repr_ & other.repr_);
   }
-  // Compute the set difference self - other
+  // Compute the set difference self - other,
+  // but ONLY for the functionality keys.
+  // Any backend bits set on self will remain unchanged.
   DispatchKeySet operator-(DispatchKeySet other) const {
-    return DispatchKeySet(repr_ & ~other.repr_);
+    return DispatchKeySet(repr_ & (full_backend_mask | ~other.repr_));
   }
+  // Perform a right bitshift
+  DispatchKeySet operator>>(uint8_t shift) const {
+    return DispatchKeySet(repr_ >> shift);
+  }
+
   // Compute self ^ other
   constexpr DispatchKeySet operator^(DispatchKeySet other) const {
     return DispatchKeySet(repr_ ^ other.repr_);
   }
-  // Perform set equality
   bool operator==(DispatchKeySet other) const {
     return repr_ == other.repr_;
   }
+  bool operator!=(DispatchKeySet other) const {
+    return repr_ != other.repr_;
+  }
   // Add a DispatchKey to the DispatchKey set.  Does NOT mutate,
   // returns the extended DispatchKeySet!
   C10_NODISCARD DispatchKeySet add(DispatchKey t) const {
     return *this | DispatchKeySet(t);
   }
-  // Remove a DispatchKey from the DispatchKey set.  This is
-  // generally not an operation you should be doing (it's
-  // used to implement operator<<)
-  C10_NODISCARD constexpr DispatchKeySet remove(DispatchKey t) const {
-    return DispatchKeySet(repr_ & ~DispatchKeySet(t).repr_);
+  C10_NODISCARD DispatchKeySet add(DispatchKeySet ks) const {
+    return *this | ks;
+  }
+
+  // Remove a DispatchKey from the DispatchKey set.
+  // Only functionality bits are allowed to be removed from a keyset.
+  // This is generally not an operation you should be doing
+  // (it's used to implement operator<<)
+  constexpr DispatchKeySet removeFunctionalityKey(DispatchKey t) const {
```

Comment:
There are some usages of DispatchKeySet::remove in functorch. Does replacing it with removeFunctionalityKey work?

Reply:
I haven't actually pushed the change yet, but just added a relevant comment here: https://github.com/pytorch/pytorch/pull/69633#issuecomment-1016855801. TLDR is that I was planning on changing `remove()` in functorch to `removeFunctionalityKey()`, but it looks like functorch tests actually all pass without me making any changes, if I keep `remove()` in.